### PR TITLE
Add function `generateSerialID`

### LINKED ISSUE #62485: A new function, `serial`
A new function, `serial`, which takes a constant string argument - a series identifier.
On every block, it goes to Keeper and atomically increments the counter with the corresponding name by the block size, then it generates and returns sequential numbers starting from the previous counter value.

The server can be configured with a keeper path - a prefix for these counter nodes, and with a limit on the total number of distinct counters (e.g., 100,000).

The usage of `serial` function will be controlled by a separate access right.

There should be a shortcut syntax for a table definition: if you specify `AUTO_INCREMENT`, it is equivalent to `DEFAULT serial('table_uuid')` or `serial('table_path_in_keeper')`. We can also support `SERIAL` for compatibility with Postgres.

As a result, it will provide a monotonic but not necessarily continuous counter. The discontinuity will appear for rejected or deduplicated INSERTs.

Alternatively, Replicated tables can store the counter inside their path in Keeper, and interpret `AUTO_INCREMENT` directly.

Caveats: special support is needed for BACKUP and RESTORE.

_Originally posted by @alexey-milovidov in https://github.com/ClickHouse/ClickHouse/issues/56228#issuecomment-1789933636_

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a new function, `generateSerialID`, which can be used to generate auto-incremental numbers in tables. Continuation of #64310 by [kazalika](https://github.com/kazalika). This closes #62485.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing)

All builds in Builds_1 and Builds_2 stages are always mandatory
and will run independently of the checks below:

- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64
- [ ]  Exclude: All with release
- [ ]  Exclude: All with debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 6acc63f12a879a6e89b3d1bec750a25f705af493 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73950/6acc63f12a879a6e89b3d1bec750a25f705af493/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/ccd0ede944185928eab30758fc390c39d3bcab2d/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Functions/generateSerialID.cpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+The server should be configured with Keeper.
+The series are stored in Keeper nodes under the path, which can be configured in `series_keeper_path` in the server configuration.
+By default, it is `/clickhouse/series/`.
+)",
+        .syntax = "generateSerialID(identifier)",
+        .arguments{
+            {"series identifier", "Series identifier (String or FixedString)"}
+        },
+        .returned_value = "Sequential numbers of type Int64 starting from the previous counter value",
```
Comment (pet74alex) [inline src/Functions/generateSerialID.cpp:166]: The actual type is UInt64

File: tests/queries/0_stateless/03129_serial_test_zookeeper.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+SELECT generateSerialID(); -- { serverError NUMBER_OF_ARGUMENTS_DOESNT_MATCH }
+SELECT generateSerialID('x', 'y'); -- { serverError NUMBER_OF_ARGUMENTS_DOESNT_MATCH }
+SELECT generateSerialID(1); -- { serverError ILLEGAL_TYPE_OF_ARGUMENT }
+
+-- Here the functions are identical and fall into common-subexpression-elimination:
+SELECT generateSerialID(currentDatabase() || 'z'), generateSerialID(currentDatabase() || 'z') FROM numbers(5);
+
+SET max_autoincrement_series = 3;
+SELECT generateSerialID('a'); -- { serverError LIMIT_EXCEEDED }
```
Comment (yakov-olkhovskiy) [inline tests/queries/0_stateless/03129_serial_test_zookeeper.sql:16]: I guess it should be `x`?
```suggestion
SELECT generateSerialID('x'); -- { serverError LIMIT_EXCEEDED }
```

File: src/Functions/generateSerialID.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    explicit FunctionSerial(ContextPtr context_) : context(context_)
+    {
+        keeper_path = context->getServerSettings()[ServerSetting::series_keeper_path];
+        zkutil::ZooKeeperPtr keeper = context->getZooKeeper();
+        size_t max_series = context->getSettingsRef()[Setting::max_autoincrement_series];
+
+        Coordination::Stat stat;
+        if (keeper->exists(keeper_path, &stat))
+        {
+            if (static_cast<size_t>(stat.numChildren) > max_series)
```
Comment (yakov-olkhovskiy) [inline src/Functions/generateSerialID.cpp:57]:
```suggestion
            if (static_cast<size_t>(stat.numChildren) >= max_series)
```

File: src/Functions/generateSerialID.cpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    explicit FunctionSerial(ContextPtr context_) : context(context_)
+    {
+        keeper_path = context->getServerSettings()[ServerSetting::series_keeper_path];
+        zkutil::ZooKeeperPtr keeper = context->getZooKeeper();
+        size_t max_series = context->getSettingsRef()[Setting::max_autoincrement_series];
+
+        Coordination::Stat stat;
+        if (keeper->exists(keeper_path, &stat))
+        {
+            if (static_cast<size_t>(stat.numChildren) >= max_series)
```
Comment (alexey-milovidov) [inline src/Functions/generateSerialID.cpp:57]: A small observation: currently it will prevent even using already created series.

**Review by yakov-olkhovskiy — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
The listing can be done using the `system.zookeeper` table, and deletion can be done with the `clickhouse-keeper-client` tool.

**Comment by davidwallacejackson** [comment]
Would it be possible to allow non-constant strings as arguments here? I'm trying to use this in a case where each partition of my data has its own ID series, to allow generating partition-specific bitmaps, and I'm not sure it's possible to do it using `generateSerialID` as it currently exists.

**Comment by alexey-milovidov** [comment]
From the architecture/internals standpoint, it is not hard to add.

**Pull request merged**